machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb

```diff
@@ -0,0 +1,535 @@
+   "outputs": [],
+   "source": [
+    "scaler = StandardScaler()\n",
+    "ddf_train[\"scaled_x\"] = scaler.fit_transform(ddf_train[[\"x\"]]).x\n",
+    "ddf_validation[\"scaled_x\"] = scaler.transform(ddf_validation[[\"x\"]]).x"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Define and initialize a Haiku neural network."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def net_function(x: jnp.ndarray) -> jnp.ndarray:\n",
+    "    net = hk.Sequential([\n",
+    "      hk.Linear(50), jax.nn.relu,\n",
+    "      hk.Linear(100), jax.nn.relu,\n",
+    "      hk.Linear(50), jax.nn.relu,\n",
+    "      hk.Linear(1),\n",
+    "    ])\n",
+    "    pred = net(x)\n",
+    "    return pred\n",
+    "\n",
+    "def initialize_net_function():\n",
+    "    net_transform = hk.transform(net_function)\n",
+    "    rng = jax.random.PRNGKey(42)\n",
+    "    num_features = 1\n",
+    "    example_x = jnp.array(np.random.random([batch_size, num_features]))\n",
+    "    params = net_transform.init(rng, example_x)\n",
+    "    return net_transform, params\n",
+    "\n",
+    "net_transform, params = initialize_net_function()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Define a mean squard error loss function."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@jax.jit\n",
+    "def loss_function(params: hk.Params, x: jnp.ndarray, y_true: jnp.ndarray) -> jnp.ndarray:    \n",
+    "    def mean_squared_error(y_true: jnp.ndarray, y_pred: jnp.ndarray) -> jnp.ndarray:\n",
+    "        loss = jnp.average((y_true - y_pred) ** 2)\n",
+    "        return loss\n",
+    "    \n",
+    "    y_pred: jnp.ndarray = net_transform.apply(params, x)\n",
+    "    loss_value: jnp.ndarray = mean_squared_error(y_true, y_pred)\n",
+    "    return loss_value"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Define and initialise an optimizer."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "optimizer: optix.InitUpdate = optix.adam(learning_rate=1e-3)\n",
+    "opt_state: optix.OptState = optimizer.init(params)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Define a predict function, and create a wrapper for use with Dask dataframe `.map_partitions()`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "@jax.jit\n",
+    "def predict(params: hk.Params, x: jnp.ndarray) -> jnp.ndarray:\n",
+    "    return net_transform.apply(params, x)\n",
+    "\n",
+    "\n",
+    "def dask_predict_wrapper(ddf: dd.DataFrame, params: hk.Params) -> jnp.ndarray:\n",
```

Comment:
```suggestion
    "def dask_predict_wrapper(df: pd.DataFrame, params: hk.Params) -> jnp.ndarray:\n",
```

Reply:
I think that you meant for this to be a pandas dataframe.  It is used in a map_partitions call.  This function will be called on each of the pandas dataframes that make up the dask dataframe.

Reply:
Good catch, thanks!